src/clr-core/internal/base/focus-trap.base.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { html, LitElement } from 'lit-element';
+import { property } from '../decorators/property';
+import { FocusTrapTracker } from '../services/focus-trap-tracker.service';
+import { addAttributeValue, isHTMLElement, removeAttributeValue } from '../utils/dom';
+import { isBrowser } from '../utils/exists';
+
+export class CdsBaseFocusTrap extends LitElement {
+  private previousFocus: Element;
+  private topReboundEl: any;
+  private bottomReboundEl: any;
+  private onFocusInEvent: any;
+
+  @property({ attribute: false })
+  private hostElement = this;
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    if (document.activeElement) {
```

Comment:
That may not work though. AoT is very finicky when it comes to this stuff. We would want to verify that approach wouldn't break server-side rendering. I know in the past I had to put these guards around each call.

Reply:
We do not need to have `isBrowser` checks within the custom elements at all, as they do not execute or get loaded on the server. The `registerElementSafely` will not attempt to register the element if not in the browser. Custom elements can not be rendered dynamically in node since the custom element registry does not exist in node but the outer HTML definition can.

Example I can render this bit of HTML in node fine

```html
<cds-modal>
  This content is SSR and is in the static HTML 
</cds-modal>
```

So for Angular Universal you must import the custom elements within the browser module only. 

Demo repo Angular Universal + Core:
https://github.com/coryrylan/core-demos/tree/master/angular-universal 

Load custom element in browser module
https://github.com/coryrylan/core-demos/blob/master/angular-universal/src/app/app.browser.module.ts#L2 // 

Use in angular component templates
https://github.com/coryrylan/core-demos/blob/master/angular-universal/src/app/app.component.html#L10 

The Vuepress docs for the new website has a similar setup

```javascript
core.js
import '@clr/core/alert';
import '@clr/core/button';
import '@clr/core/icon';
import { ClarityIcons, homeIcon, angleDoubleIcon, angleIcon} from '@clr/core/icon-shapes';

ClarityIcons.addIcons(angleDoubleIcon, angleIcon, homeIcon);
``` 

```javascript
// enhanceApp.js 
import './styles/index.scss';

export default ({
  Vue, // the version of Vue being used in the VuePress app
  options, // the options for the root Vue instance
  router, // the router instance for the app
  siteData, // site metadata,
  isServer
}) => {
  if (!isServer) {
    import('./core');
  }
}
```

You can render custom elements dynamically in node with additional setup on the server side either using a custom elements node shim like [JSDom](https://github.com/jsdom/jsdom)  which recently added [custom element support](https://github.com/jsdom/jsdom/pull/2548).

Or you can use a prerender with something like [puppeteer](https://developers.google.com/web/tools/puppeteer). This is how the Angular static site renderer Scully works https://github.com/scullyio/scully.

Reply:
I see. As you mentioned, `registerElementSafely` is not attempting to work during server-side rendering. It doesn't look like it's chocking during the pre-rendering process. So in this case, do we need to be checking if it's running in a browser?

Reply:
For pre-rendering using with Scully or Puppeteer it works without checks since it uses a headless browser. Pre-rendering with Angular Universal or anything that renders to a string follows the same pattern as server side rendering. It must load the custom elements only in the browser.

The Universal demo I linked it could be updated to prerender. The code of where the custom elements are imported would stay the same. So in the pre-render scenarios we would not need to check if we are running in the browser.